spec/controllers/admin/tests_controller_spec.rb

```diff
@@ -0,0 +1,275 @@
+      end
+    end
+  end
+
+  describe "PATCH #update" do
+    context "with valid parameters" do
+      let(:new_attributes) { { name: "Updated Test Name" } }
+
+      before { patch :update, params: { id: test_obj.id, test: new_attributes } }
+
+      it "updates the test" do
+        test_obj.reload
+        expect(test_obj.name).to eq("Updated Test Name")
+      end
+
+      it "assigns @test" do
+        expect(assigns(:test)).to eq(test_obj)
+      end
+
+      it "sets success flash message" do
+        expect(flash[:success]).to be_present
+      end
+
+      it "redirects to test show page" do
+        expect(response).to redirect_to(admin_test_path(test_obj))
+      end
+    end
+
+    context "with invalid parameters" do
+      before { patch :update, params: { id: test_obj.id, test: invalid_attributes } }
+
+      it "does not update the test" do
+        original_name = test_obj.name
+        test_obj.reload
+        expect(test_obj.name).to eq(original_name)
+      end
+
+      it "assigns @test with errors" do
+        expect(assigns(:test)).to eq(test_obj)
+        expect(assigns(:test).errors).not_to be_empty
+      end
+
+      it "sets danger flash message" do
+        expect(flash[:danger]).to be_present
+      end
+
+      it "renders edit template" do
+        expect(response).to render_template(:edit)
+      end
+
+      it "returns unprocessable entity status" do
+        expect(response).to have_http_status(:unprocessable_entity)
+      end
+    end
+
+    context "when test does not exist" do
+      before { patch :update, params: { id: 999999, test: valid_attributes } }
+
+      it "sets danger flash message" do
+        expect(flash[:danger]).to be_present
+      end
+
+      it "redirects to tests index" do
+        expect(response).to redirect_to(admin_tests_path)
+      end
+    end
+  end
+
+  describe "DELETE #destroy" do
+    context "when test can be destroyed" do
+      before { delete :destroy, params: { id: test_obj.id } }
+
+      it "destroys the test" do
+        expect(Test.exists?(test_obj.id)).to be_falsey
+      end
+
+      it "sets success flash message" do
+        expect(flash[:success]).to be_present
+      end
+
+      it "redirects to tests index" do
+        expect(response).to redirect_to(admin_tests_path)
+      end
+    end
+
+    context "when test cannot be destroyed" do
+      before do
+        allow_any_instance_of(Test).to receive(:destroy).and_return(false)
+        delete :destroy, params: { id: test_obj.id }
+      end
+
+      it "does not destroy the test" do
+        expect(Test.exists?(test_obj.id)).to be_truthy
+      end
+
+      it "sets danger flash message" do
+        expect(flash[:danger]).to be_present
+      end
+
+      it "redirects to test show page" do
+        expect(response).to redirect_to(admin_test_path(test_obj))
+      end
+    end
+
+    context "when test does not exist" do
+      before { delete :destroy, params: { id: 999999 } }
```

Comment:
```suggestion
      before { delete :destroy, params: { id: -1 } }
```
Tương tự cho những chỗ khác